### ISSUE TITLE
Surprising syntax rule about invoking a delegate defined as a property.

### ISSUE THREAD

**Issue by ericmutta** [dotnet/vblang#243]
Consider the following:

```vb
  Public Class CTest
    Public Action1 As Action(Of Action)

    Public Property Action2 As Action(Of Action)

    Public Function Action3() As Action(Of Action)
    End Function

    Public Sub Test()
      'this is OK. Action1 is just a field.
      Me.Action1(Sub() Console.WriteLine("blah"))

      'this FAILS. Action2 has same type as Action1 but is now a property.
      'Compiler says: TOO MANY ARGUMENTS TO Action2
      Me.Action2(Sub() Console.WriteLine("blah"))

      'this is WHAT THE COMPILER WANTS (notice the double brackets after Action2).
      Me.Action2()(Sub() Console.WriteLine("blah"))

      'I expect to use brackets here because Action3 is a function.
      Me.Action3()(Sub() Console.WriteLine("blah"))

      Me.Action2 = Nothing '<--- this I expect and do all the time.
      Me.Action2() = Nothing '<--- this works but is just weird!
    End Sub
  End Class
```

The comments spell it out but the special point of interest is the line:

```vb
      Me.Action2(Sub() Console.WriteLine("blah"))
```

The `Action1` field and the `Action2` property have identical types (i.e  `Action(Of Action)`). I was initially using the field form in my code which allows me to invoke the delegate as follows:

```vb
      Me.Action1(Sub() Console.WriteLine("blah"))
```

...but later I decided to use a property, so I added the `Property` keyword (resulting in a declaration similar to `Action2`), but now all the call sites invoking the delegate were broken and the compiler wanted code like this:

```vb
      Me.Action2(Sub() Console.WriteLine("blah")) '<-- CALL THIS FORMAT 1
```

...to be written like this (i.e put empty brackets immediately after the property name):

```vb
      Me.Action2()(Sub() Console.WriteLine("blah")) '<-- CALL THIS FORMAT 2
```

Is there any reason that format 1 above can't work until the compiler requires format 2?

Also how is this legal syntax:

```vb
      Me.Action2() = Nothing '<--- this works but is just weird!
```
...? It looks like calling a function and then nulling its result (it's impossible to write a function that would allow this today without VB support for [ref returns](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/ref-returns)).

CC: @AnthonyDGreen

**Comment by franzalex** [comment]
What you observe on the property is indeed the expected behaviour. I remember coming across an article¹ that stated something I'll paraphrase as:

> Properties are wrappers around functions for retrieving and setting fields.

It went on to state that a class like _Code Block 1_ gets compiled as in _Code Block 2_

**Code Block 1**
``` vb.net
Public Class Test1
    Private _prop1 As String
    Public Property Prop1 As String
        Get
            Return _prop1
        End Get
        Set(value As String)
            _prop1 = value
        End Set
    End Property
End Class
```

**Code Block 2**
``` vb.net
Public Class Test1
    Private _prop1 As String
    Public Function Get_Prop1() As String
        Return _prop1
    End Function
    Public Sub Set_Prop1(value As String)
            _prop1 = value
    End Sub
End Class
```

By this, it can be seen why the procedure for invoking property and method delegates are identical. I see no way this can be changed except a reimplementation of properties across both C# and VB.NET.

———
¹ I unfortunately don't remember the title or I'd have given a reference

**Comment by ericmutta** [comment]
> @franzalex I see no way this can be changed except a reimplementation of properties across both C# and VB.NET.

I understand that properties are implemented as functions underneath and the compiler hides away those details, making them "feel" like regular fields in the two main scenarios: reading a property and assigning a property. However **this affordance does not extend to the invocation syntax** which works for delegate-typed fields but breaks for delegate-typed properties, My question is...why does it break?

The question is significant because it affects code maintainenance. In my case I had a delegate-typed field and later wanted it to be a property (a somewhat typical refactoring I would say). Changing it to a property broke all code that was invoking the delegate because the compiler wanted extra brackets after the delegate-typed property name. This a small syntax requirement that should not require any major changes within/across languages.

PS: VB happens to be quite flexible with its rules for using brackets during invocation. For example you can write `SomeFunction` instead of `SomeFunction()` to invoke a function which takes no parameters. Omitting the brackets makes the function look like it is field/property. Likewise you can write `Dim x = Date.Now()` which makes `Now` look like a function when in fact it is a [property](https://msdn.microsoft.com/en-us/library/system.datetime.now(v=vs.110).aspx). In other words, VB doesn't seem to care much whether you include/exclude the brackets and always does what you expect...but not if you are trying to invoke a delegate-typed property. Seems inconsistent to me for no apparent reason I can think of :man_shrugging:

**Comment by AnthonyDGreen** [comment]
It's because VB (and COM) have always had parameterized properties. It's technically more correct to say that VB allows you to omit the parameter list when invoking a parameterless property. So when you add the parentheses it's just writing explicitly what the code always means. If I'm not mistaken the rule is the same for function invocations, which may omit parentheses when unnecessary. I once theorized that this was valuable because without the rule that lets you omit the parentheses whenever you wanted to index an object returned from a property you'd have to write this:

``` VB.NET
' First parens for property, second parents for indexer of ColumnCollection object.
dataset.Columns()(0).Name = ""
```

Fields are different.

You could special case this scenario:

``` VB.NET
Me.Action()
```

Because it's not valid to just invoke a property as a statement. But then it would be inconsistent with this case:

``` VB.NET
Console.WriteLine(Me.Action())
```

Does that simply write the ToString() value of the delegate returned from the property or does it invoke the delegate and write its return value?

It does work intuitively for the read case if the delegate takes any arguments:

``` VB.NET
Module Module1
    Sub Main()
        Dim c = New C
        c.P = Function(left, right) left + right

        Console.WriteLine(c.P(2, 3))

    End Sub
End Module

Class C
    Property P As Func(Of Integer, Integer, Integer)
End Class
```

That works in an expression context but not a statement one. An argument could be made that that should be valid.

**Comment by ericmutta** [comment]
@AnthonyDGreen thanks for responding at length (especially the COM history...it's been years since I thought about COM!).

Let me try again with another simpler example:

```vb
Public Module Module1
  Public Class Foo
    Public Property Run As Action(Of String)
  End Class

  Public Sub Main()
    Dim f = New Foo
    f.Run("hello!") '<-- this DOES NOT work.

    Dim r = f.Run
    r("hello!") '<-- but this DOES WORK.
  End Sub
End Module
```

**Could the language rules be improved so that one can directly invoke delegates THAT TAKE ONE OR MORE PARAMETERS and are stored in a property, in the same way we can do so when such delegates are stored in fields or local variables**? In other words, when the compiler sees this (which is NOT allowed today):

```vb
SomeObject.SomeDelegateTypedProperty(foo, bar, baz)
```

...it SHOULD treat it as if it was written like this (which is the transformation I did by hand when the compiler complained about the above code):

```vb
Dim TheDelegate = SomeObject.SomeDelegateTypedProperty
TheDelegate(foo, bar, baz)
```

At the moment, in the code shown earlier in this comment, the compiler produces this error message:

![vs_property_invocation](https://user-images.githubusercontent.com/20465797/35101668-66ef2b18-fc71-11e7-8269-d8519295f76d.png)

...and wants you to write `f.Run()("hello!")`...i.e **you are forced to include the brackets, contrary to the rule that says you can omit them when a property has no parameters**.

To reiterate: this is a problem because if the symbol `f.Run` resolves to a delegate-typed field then ` f.Run("hello!")` is perfectly valid. However if you change `f.Run` to a property,  you have to **go to all call sites and add the brackets** so it reads `f.Run()("hello!")`. If you start with a property and switch to a field, you have to **go to all call sites and remove the brackets**.

Since (1) delegates are pretty much inevitable in any non-trivial code base and (2) switching from a field to a property is both a common and advisable refactoring, I would like to make the case that fixing this issue would be a small syntactic change with real customer value 👍 👍 

PS: added some corporate-speak at the end there for maximum dramatic effect 😄

**Comment by bandleader** [comment]
@ericmutta Firstly, as I understand, this isn't because of properties or delegates with a delegate parameter; you can demonstrate the problem with a simple function, like this:
```vb
        Sub Test()
            ReturnADelegate(4) 'Error: Too many arguments to Function ReturnADelegate() As Action(Of Integer).
            ReturnADelegate()(4) 'Correct, works
            ReturnADelegate.Invoke(4) 'Also correct and works
        End Sub

        Function ReturnADelegate() As Action(Of Integer)
        End Function
```

Secondly, I'm not sure anything needs to be changed here. Since properties and functions are argumentful, if you put a set of parentheses after a property/function name, it is assumed to be the argument list for that property/function.

Yes, VB lets you omit empty argument lists; however here you *aren't omitting it -- you're supplying an argument list with a different arity,* hence the "too many arguments" error.

If you think the double parens are hard to read, you can use (as above) the real delegate invocation syntax `Test.Invoke(...)`.

**As regards your need to refactor from a field to a property:** you should use the `.Invoke` syntax (which is the real syntax anyway) to eliminate the ambiguity, so you'll be covered.

P.S. The compiler could _technically_ be changed to realize that the first function call is argumentless, and therefore pass the argument list to pass to `Invoke` instead. But that is inherently ambiguous, and could cause worse issues than you are having: if a parameterless property is refactored with an optional argument, the parameter list will then suddenly be used for the property, and not for the delegate that it returns. The current way is much more sensible IMHO.

**Comment by bandleader** [comment]
@AnthonyDGreen Contrary to what I wrote above, the syntax @ericmutta is requesting curiously *does* work in expressions, just not in invocations. This is most apparent with arrays/functions, which can be used in expressions. Consider the below code:
```vb
        Sub Test2()
            ReturnADelegate(1) 'Error: Too many arguments to Function ReturnADelegate() As Action(Of Integer).
            Dim t1 = ReturnADelegate(1) 'Error should be the same as above, instead it is simply "Expression does not produce a value."

            'Following are examples with arrays/functions, and where it actually works:

            Dim t2 = ReturnAnArray(1) 'This compiles just fine! Returns "Zero"
            ReturnAnArray(1) 'Error: Too many arguments to Function ReturnAnArray() As String().

            Dim t3 = ReturnAFunc(1) 'This works too! Returns 2
            ReturnAFunc(1) 'Error: Too many arguments to Function ReturnAFunc() As Func(Of Integer, Integer).
        End Sub

        'All these functions are PARAMETERLESS but return an object that can take parameters/indexes
        Function ReturnAnArray() As String()
            Return {"Zero", "One"}
        End Function
        Function ReturnAFunc() As Func(Of Integer, Integer)
            Return Function(i) i + 1
        End Function
        Function ReturnADelegate() As Action(Of Integer)
            Return Sub(i) Debug.Print("Done " & i)
        End Function
``` 

Why would that be, and is it a bug?

**Comment by reduckted** [comment]
@bandleader That's very interesting.

_If_ this were to be fixed it would be a breaking change. @KathleenDollard, do you know if compiler flags to enable stricter compilation have ever been considered in VB? Something like what the [TypeScript compiler](http://www.typescriptlang.org/docs/handbook/compiler-options.html) has comes to mind. When a new version introduces stricter type checks, the compiler gets a new flag so that it can remain backward compatible. That sort of thing seems like it would help tighten up these strange behaviours, while being able to remain backward compatible.

FYI, for anyone who hasn't heard the news: https://gist.github.com/AnthonyDGreen/d027ebb7052cce2400b670c7d4262014

**Comment by KathleenDollard** [comment]
Adding 

@gafter and @AlekseyTs for history. 

I want to avoid compiler flags due to the complexity it would cause in both the compiler itself and user comprehension of code. I think you should look at code and be able to predict behavior with as little environmental knowledge as possible.

**Comment by bandleader** [comment]
@reduckted @KathleenDollard With regard to compiler options, I think the VB way is more with `Option Foo On` at the top of the .vb file. (Though again, I don't think this issue is a candidate for that.)

@KathleenDollard Great to hear about your appointment -- good luck with everything, and thank you in advance for your work on VB! With a language as great as VB comes great responsibility...

**Comment by ericmutta** [comment]
> @bandleader ...the syntax @ericmutta is requesting curiously does work in expressions, just not in invocations.

That is quite interesting, good catch!

> @reduckted FYI, for anyone who hasn't heard the news: 

Thanks for sharing that, I would have completely missed it then started grumbling that Athony isn't answering us any more! 

PS: @KathleenDollard, congrats on your new appointment as Supreme Leader in all things VB 😃

**Comment by ericmutta** [comment]
> @KathleenDollard I want to avoid compiler flags due to the complexity it would cause in both the compiler itself and user comprehension of code.

Totally agree with this. I like how there's only one version of VB (the one that matches whatever version of Visual Studio you happen to be using) and that once you've set Option Explicit/Infer/Strict/Compare appropriately (they are the same for ALL my projects) you never have to think about what the language will do 👍 👍

**Comment by AnthonyDGreen** [comment]
I'm not dead ;)

I agree with Kathleen about a compiler flag being the wrong thing here. It's also not really needed. The compiler does the correct thing for delegates of 1 or more parameter in an expression context. It can be made to use the same logic in the statement context. There's actually a small cluster of cases where invocations in a statement context don't work in VB today, including late-bound. In 2009 I noticed that `DynamicObject.TryInvoke` can only be called from VB in an expression context:

``` VB.NET
Imports System.Dynamic

Module Module1
    Sub Main()
        Dim lateBound As Object = New MyConsoleWriter

        ' Compiles and correctly invokes the object late-bound.
        Console.WriteLine(lateBound(Date.Now))

        ' Won't compile, only invocable in an expression context.
        lateBound(Date.Now)
    End Sub
End Module

Class MyConsoleWriter
    Inherits DynamicObject

    Public Overrides Function TryInvoke(binder As InvokeBinder, args As Object(), ByRef result As Object) As Boolean
        args = If(args, {})
        If args.Length = 1 Then
            Console.WriteLine(args(0).GetType())
            result = args(0)
            Return True
        End If

        Return MyBase.TryInvoke(binder, args, result)
    End Function
End Class
```

The compiler should be able to invoke dynamic objects _and_ delegates just as easily as a statement as it does as an expression, imo. Not sure if there is a single fix for all cases though.

**Comment by bandleader** [comment]
@AnthonyDGreen In this case, I actually opine that the 'bug' is the fact that the expression syntax allows it! The parentheses (argument list) are ambiguous, as detailed above (they could suddenly be passed to the delegate instead if the delegate got a method). 

Omitting an empty argument list is a classic and great VB readability feature, but if an argument list **IS** given, it has to be used, unambiguously, for the method name that precedes it. 

Passing it to the *return result* of the method **should** require `delegate()(5)` IMHO. This doesn't *ruin* clarity/readability, it *enhances* it by removing ambiguity.

I'm not sure if we can fix this anymore, as it's a (rare but) breaking change.

**Comment by AnthonyDGreen** [comment]
@bandleader absolutely cannot change this--requiring the extra set of parenthesis--many Bothans would die.

**Comment by paul1956** [comment]
@ericmutta I wish it were true that the one version of VB matched VS, The correct statement is one default version of VB for each major release of Visual Studio. With new features introduced in 15.3 and 15.5 there are 3 but VS always defaults to 15.0 and you get messages to upgrade when you try to use new features (sorry if this sounds like ranting not my intent).

**Comment by KathleenDollard** [comment]
@ericmutta you are correct that the default is 15.0. FWIW, C# defaults to "latest major" so this is consistent. 

The gap is that if you want a non-default policy, in C# you have a drop down (in an obscure location) and in VB, you edit the vbproj.

**Comment by paul1956** [comment]
@KathleenDollard I don't do much C# but I thought I got a C# CodeFix for it AND the error message gives a link to detailed instructions (same link for VB) but the instructions only work for C#. I would be thrilled if it was as easy in VB as C# or even if the instructions just worked. I have reported other cases where a VS error is the same for VB and C# but fix only works for C#.
![image](https://user-images.githubusercontent.com/12789369/43619416-816ecb28-9682-11e8-9138-02b4a46fba92.png).
